### PR TITLE
fix #9872: setLen now works properly at CT

### LINKED ISSUE #9865: compiler SIGSEGV when calling setLen on `seq[tuple]` or `seq[object]` in macro
/cc @Araq 
while trying to further reduce https://github.com/nim-lang/Nim/issues/9864 I ran into another bug:

```nim

#[
KEY SIGSEGV
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
D20181204T222045

Traceback (most recent call last)
/Users/timothee/git_clone/nim/Nim/compiler/nim.nim(109) nim
/Users/timothee/git_clone/nim/Nim/compiler/nim.nim(73) handleCmdLine
/Users/timothee/git_clone/nim/Nim/compiler/cmdlinehelper.nim(91) loadConfigsAndRunMainCommand
/Users/timothee/git_clone/nim/Nim/compiler/main.nim(168) mainCommand
/Users/timothee/git_clone/nim/Nim/compiler/main.nim(78) commandCompileToC
/Users/timothee/git_clone/nim/Nim/compiler/modules.nim(134) compileProject
/Users/timothee/git_clone/nim/Nim/compiler/modules.nim(79) compileModule
/Users/timothee/git_clone/nim/Nim/compiler/passes.nim(194) processModule
/Users/timothee/git_clone/nim/Nim/compiler/passes.nim(86) processTopLevelStmt
/Users/timothee/git_clone/nim/Nim/compiler/cgen.nim(1487) myProcess
/Users/timothee/git_clone/nim/Nim/compiler/ccgstmts.nim(1186) genStmts
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2362) expr
/Users/timothee/git_clone/nim/Nim/compiler/ccgcalls.nim(552) genCall
/Users/timothee/git_clone/nim/Nim/compiler/ccgcalls.nim(191) genPrefixCall
/Users/timothee/git_clone/nim/Nim/compiler/ccgcalls.nim(174) genArg
/Users/timothee/git_clone/nim/Nim/compiler/cgen.nim(509) initLocExprSingleUse
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2388) expr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(1277) genObjConstr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(1267) handleConstExpr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2629) genConstExpr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2562) genConstObjConstr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2550) getNullValueAuxT
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2517) getNullValueAux
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2529) getNullValueAux
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2610) genConstExpr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2583) genConstSeq
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2629) genConstExpr
/Users/timothee/git_clone/nim/Nim/compiler/ccgexprs.nim(2557) genConstObjConstr
/Users/timothee/git_clone/nim/Nim/compiler/ast.nim(1415) skipTypes
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
]#

import macros

type
  # Bar = tuple[key: int] # also fails
  Bar = object

  Foo* = object
    data: seq[Bar]

proc bar(a: Foo)=discard

macro dispatchGen(): untyped =
  var shOpt: Foo
  shOpt.data.setLen 1

  result = quote do:
     bar(`shOpt`)

dispatchGen()
```

## links
* not sure whether related: https://github.com/nim-lang/Nim/issues/5870 (stracktrace has some similarity) but in my case code seems legal
* maybe related: https://github.com/nim-lang/Nim/issues/9872 (but stacktrace is different, and doesn't use quote do)

Note: marking high priority as I don't know a good workaround
(newSeq isn't always possible when we need to change the length after the seq is created)

### LINKED ISSUE #9872: SIGSEGV: `setLen` on a seq doesn't construct objects at CT
/cc @Araq 
```nim
#[
KEY ICE crash

/Users/timothee/git_clone/nim/timn/config.nims [config.nims used]
/Users/timothee/git_clone/nim/Nim/compiler/nim.nim(109) nim
/Users/timothee/git_clone/nim/Nim/compiler/nim.nim(73) handleCmdLine
/Users/timothee/git_clone/nim/Nim/compiler/cmdlinehelper.nim(91) loadConfigsAndRunMainCommand
/Users/timothee/git_clone/nim/Nim/compiler/main.nim(168) mainCommand
/Users/timothee/git_clone/nim/Nim/compiler/main.nim(78) commandCompileToC
/Users/timothee/git_clone/nim/Nim/compiler/modules.nim(134) compileProject
/Users/timothee/git_clone/nim/Nim/compiler/modules.nim(79) compileModule
/Users/timothee/git_clone/nim/Nim/compiler/passes.nim(194) processModule
/Users/timothee/git_clone/nim/Nim/compiler/passes.nim(86) processTopLevelStmt
/Users/timothee/git_clone/nim/Nim/compiler/sem.nim(603) myProcess
/Users/timothee/git_clone/nim/Nim/compiler/sem.nim(571) semStmtAndGenerateGenerics
/Users/timothee/git_clone/nim/Nim/compiler/semstmts.nim(1987) semStmt
/Users/timothee/git_clone/nim/Nim/compiler/semexprs.nim(915) semExprNoType
/Users/timothee/git_clone/nim/Nim/compiler/semexprs.nim(2548) semExpr
/Users/timothee/git_clone/nim/Nim/compiler/semstmts.nim(1927) semStmtList
/Users/timothee/git_clone/nim/Nim/compiler/semexprs.nim(2436) semExpr
/Users/timothee/git_clone/nim/Nim/compiler/semexprs.nim(897) semDirectOp
/Users/timothee/git_clone/nim/Nim/compiler/semexprs.nim(790) afterCallActions
/Users/timothee/git_clone/nim/Nim/compiler/sem.nim(466) semMacroExpr
/Users/timothee/git_clone/nim/Nim/compiler/vm.nim(1999) evalMacroCall
/Users/timothee/git_clone/nim/Nim/compiler/vm.nim(580) rawExecute
/Users/timothee/git_clone/nim/Nim/lib/system.nim(2969) sysFatal
Error: unhandled exception: index out of bounds [IndexError]
]#

type Foo = object
  index: int

macro fun(): untyped =
  let n = 1
  var foos: seq[Foo]
  foos.setLen n
  for i in 0 ..< n:
    discard foos[i] # works
    echo foos[i].index # Error: unhandled exception: index out of bounds [IndexError]

fun()
```

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @Araq 
* fix #9872: setLen now works properly at CT
* fix #9865

### CODE REVIEW

File: tests/vm/tsetlen.nim [hunk 1]
```diff
@@ -0,0 +1,32 @@
+type Foo = object
+  index: int
+
+block:
+  proc fun[T]():auto =
```
Comment (Araq) [inline tests/vm/tsetlen.nim]: Space after the colon. Unclear, why it has a return type.

File: tests/vm/tsetlen.nim [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      let temp = foos[i]
+      when T is object:
+        doAssert temp.index == 0
+      when T is ref object:
+        doAssert temp == nil
+      doAssert temp == foo
+    0
+
+  static:
+    discard fun[Foo]()
```
Comment (Araq) [inline tests/vm/tsetlen.nim]: Without a return type you also don't need a ``discard``.

**Comment by timotheecour** [comment]
PTAL

**Comment by Araq** [comment]
Can be merged with ``[backport]`` once the CIs are green.

**Comment by timotheecour** [comment]
can I merge it myself in case it's green ? if so, shall i rebase and change commit msg to `fix #9872: setLen now works properly at CT [backport]` ?

**Comment by Araq** [comment]
> can I merge it myself in case it's green ? if so, shall i rebase and change commit msg to fix #9872: setLen now works properly at CT [backport] ?

Yes and yes.

**Comment by timotheecour** [comment]
ok, I re-pushed to do these changes and added a `merge_when_passes_CI` label (see https://github.com/nim-lang/Nim/issues/9847) and will merge as soon as green again

**Pull request merged**